builder/dockerfile/copy.go

```diff
@@ -288,33 +289,56 @@ func walkSource(source builder.Source, origPath string) ([]string, error) {
 	return subfiles, nil
 }
 
-type sourceDownloader func(string) (builder.Source, string, error)
+type sourceDownloader func(url, dest string) (builder.Source, string, error)
 
 func newRemoteSourceDownloader(output, stdout io.Writer) sourceDownloader {
-	return func(url string) (builder.Source, string, error) {
-		return downloadSource(output, stdout, url)
+	return func(url, dest string) (builder.Source, string, error) {
+		return downloadSource(output, stdout, url, dest)
 	}
 }
 
-func errOnSourceDownload(_ string) (builder.Source, string, error) {
+func errOnSourceDownload(_, _ string) (builder.Source, string, error) {
 	return nil, "", errors.New("source can't be a URL for COPY")
 }
 
-func downloadSource(output io.Writer, stdout io.Writer, srcURL string) (remote builder.Source, p string, err error) {
+func downloadSource(output io.Writer, stdout io.Writer, srcURL, dest string) (remote builder.Source, p string, err error) {
 	u, err := url.Parse(srcURL)
 	if err != nil {
 		return
 	}
-	filename := filepath.Base(filepath.FromSlash(u.Path)) // Ensure in platform semantics
-	if filename == "" {
-		err = errors.Errorf("cannot determine filename from url: %s", u)
-		return
+	var filename string
+	if u.Path != "" && !strings.HasSuffix(u.Path, "/") {
+		filename = filepath.Base(filepath.FromSlash(u.Path)) // Ensure in platform semantics
+		if filename == "" {
+			err = errors.Errorf("cannot determine filename from url: %s", u)
+			return
+		}
+	} else if dest != "" && !strings.HasSuffix(dest, "/") {
+		filename = filepath.Base(filepath.FromSlash(dest))
+		if filename == "" {
+			err = errors.Errorf("cannot determine filename from dest: %s", dest)
+			return
+		}
 	}
 
 	resp, err := remotecontext.GetWithStatusError(srcURL)
 	if err != nil {
 		return
 	}
+	if filename == "" {
+		// Guess filename based on Content-Disposition
+		if contentDisposition := resp.Header.Get("Content-Disposition"); contentDisposition != "" {
+			if _, params, err := mime.ParseMediaType(contentDisposition); err == nil {
+				if params["filename"] != "" && !strings.HasSuffix(params["filename"], "/") {
+					filename = params["filename"]
+				}
+			}
+		}
+		if filename == "" {
```

Comment:
Make sure to sanitise the filename we received from the `"Content-Disposition"` header; we're only interested in a filename, not any (if passed) path information